File: chapter_5/binarySearch.cpp
#include <iostream>
#include <string>

int binarySearch(int list[], int item, int left, int right);


int main() {
    
    int list[] = {1, 4, 8, 11, 13, 16, 19, 19};
    int size = sizeof(list) / sizeof(int);
    int item = 13;
    
    int index = binarySearch(list, item, 0, size-1);
    std::cout << "Index is " << index << std::endl;
    return 0;
}


int binarySearch(int list[], int item, int left, int right) {
    
    if (left > right) 
        return -1;
    
    // Integer div rounds down.
    int ret;
    int mid = (left + right) / 2;

    if (list[mid] == item) { 
        ret = mid;
    } else if (list[mid] < item) {
        ret = binarySearch(list, item, mid+1, right);
    } else if (list[mid] > item) {
        ret = binarySearch(list, item, left, mid-1);
    }

    return ret;
}

File: chapter_5/binarySearch.py
def binarySearch(item, items, left=None, right=None):

    print(f"Searching: {items[left:right+1]}")

    if left > right:
        return -1

    mid = (left + right) // 2
    if items[mid] == item:
        return mid
    if items[mid] > item:
        return binarySearch(item, items, left, mid - 1)
    elif items[mid] < item:
        return binarySearch(item, items, mid + 1, right)


if __name__ == "__main__":
    item = 13
    items = [1, 4, 8, 11, 13, 16, 19, 19]
    print(binarySearch(item, items, 0, len(items) - 1))
